File: util/src/com/syrus/io/BellcoreStructure.java
/*
 * $Id: BellcoreStructure.java,v 1.16 2005/06/17 07:12:06 saa Exp $
 *
 * Copyright © 2004 Syrus Systems.
 * Dept. of Science & Technology.
 * Project: AMFICOM.
 */

package com.syrus.io;

import java.util.Date;

/**
 * @version $Revision: 1.16 $, $Date: 2005/06/17 07:12:06 $
 * @author $Author: saa $
 * @module util
 */
public class BellcoreStructure {
	protected static final String FIELD_NAME_MAP = "Map";
	protected static final String FIELD_NAME_GENPARAMS = "GenParams";
	protected static final String FIELD_NAME_SUPPARAMS = "SupParams";
	protected static final String FIELD_NAME_FXDPARAMS = "FxdParams";
	protected static final String FIELD_NAME_KEYEVENTS = "KeyEvents";
	protected static final String FIELD_NAME_LNKPARAMS = "LnkParams";
	protected static final String FIELD_NAME_DATAPTS = "DataPts";
	protected static final String FIELD_NAME_CKSUM = "Cksum";
	protected static final String FIELD_NAME_HP_MINI_SPECIAL = "HPMiniSpecial";

	static final int MAP = 1;
	static final int GENPARAMS = 2;
	static final int SUPPARAMS = 3;
	static final int FXDPARAMS = 4;
	static final int KEYEVENTS = 5;
	static final int LNKPARAMS = 6;
	static final int DATAPOINTS = 7;
	static final int SPECIAL = 8;
	static final int CKSUM = 9;

	//флаг, показывающий есть ли в файле определенное поле
	private boolean hasMap = false;
	private boolean hasGen = false;
	private boolean hasSup = false;
	private boolean hasFxd = false;
	private boolean hasKey = false;
	private boolean hasLnk = false;
	private boolean hasData = false;
	/**
	 * @todo Field is never read locally.
	 */
	private boolean hasSpecial = false;
	private int blocks = 0; // общее число полей
	int specials = 0; // число полей specials (может быть произвольным)

	// экземпляры классов, представляющих собой поля данных в формате bellcore
	Map map;
	GenParams genParams;
	SupParams supParams;
	FxdParams fxdParams;
	KeyEvents keyEvents;
	LnkParams lnkParams;
	DataPts dataPts;
	Cksum cksum;
	Special[] special;

	public String title = "";
	public String schemePathId;
	public String measurementId;
	public String monitoredElementId;

	void addField(int type) {
		switch (type) {
			case MAP: // required
				if (!this.hasMap) {
					this.map = new Map();
					this.hasMap = true;
					this.blocks++;
				}
				break;
			case GENPARAMS: // required
				if (!this.hasGen) {
					this.genParams = new GenParams();
					this.hasGen = true;
					this.blocks++;
				}
				break;
			case SUPPARAMS: // required
				if (!this.hasSup) {
					this.supParams = new SupParams();
					this.hasSup = true;
					this.blocks++;
				}
				break;
			case FXDPARAMS: // required
				if (!this.hasFxd) {
					this.fxdParams = new FxdParams();
					this.hasFxd = true;
					this.blocks++;
				}
				break;
			case KEYEVENTS:
				if (!this.hasKey) {
					this.keyEvents = new KeyEvents();
					this.hasKey = true;
					this.blocks++;
				}
				break;
			case LNKPARAMS:
				if (!this.hasLnk) {
					this.lnkParams = new LnkParams();
					this.hasLnk = true;
					this.blocks++;
				}
				break;
			case DATAPOINTS:
				if (!this.hasData) {
					this.dataPts = new DataPts();
					this.hasData = true;
					this.blocks++;
				}
				break;
			case SPECIAL:
				this.special[this.specials++] = new Special();
				this.hasSpecial = true;
				this.blocks++;
				break;
			case CKSUM: // required
				if (!this.hasMap) {
					this.cksum = new Cksum();
					this.blocks++;
				}
		}
	}

	// ------------------ Map -------------------//
	class Map {
		public int mrn = 0; // Map Revision Number
		public int mbs = 0; // Map Block Size (in bytes)
		public short nb = 1; // Number of Blocks (including Map)
		// Block Info:
		public String[] bId; // Block ID
		public int[] bRev; // Block Revision Number
		public int[] bSize; // Block Size (in bytes)

		public int getSize() {
			int size = 8;
			for (int i = 1; i < this.nb; i++)
				size = size + this.bId[i].length() + 7;
			return size;
		}
	}

	// ----------- General Parameters ------------//
	class GenParams {
		public String lc = "EN"; // Language Code (2 bytes)
		public String cid = " "; // Cable ID
		public String fid = " "; // Fiber ID
		public short ft = 652; // Fiber Type
		public short nw = 1310; // Nominal Wavelength
		public String ol = " "; // Originating Location
		public String tl = " "; // Terminating Location
		public String ccd = " "; // Cable Code
		public String cdf = "BC"; // Current Data Flag (2 bytes)
		public int uo = 0; // User Offset
		public int uod = 0; // User Offset Distance
		public String op = " "; // Operator
		public String cmt = " "; // Comment

		public int getSize() {
			return (10
					+ 7
					+ this.cid.length()
					+ this.fid.length()
					+ this.ol.length()
					+ this.tl.length()
					//+ 6
					+ this.ccd.length()
					+ this.op.length() + this.cmt.length());
		}
	}

	// ----------- Supplier Parameters ------------//
	class SupParams {
		public String sn = " "; // Supplier Name
		public String mfid = " "; // OTDR Mainframe ID
		public String otdr = " "; // OTDR Mainframe SerNum
		public String omid = " "; // Optical Module ID
		public String omsn = " "; // Optical Module SerNum
		public String sr = " "; // Software Revision
		public String ot = " "; // Other

		public int getSize() {
			return (7
					+ this.sn.length()
					+ this.mfid.length()
					+ this.otdr.length()
					+ this.omid.length()
					+ this.omsn.length()
					+ this.sr.length() + this.ot.length());
		}
	}

	// ----------- Fixed Parameters ------------//
	class FxdParams {
		public long dts = 0; // Date/Time Stamp (in ms)
		public String ud = "km"; // Units of Distanse (2 bytes)
		public short aw = 13100; // Actual Wavelength
		public int ao = 0; // Acquision Offset
		public int aod = 0; // Acquision Offset Distance
		public short tpw = 0; // Total Number of Pulse Width Used
		public short[] pwu; // Pulse Width Used
		public int[] ds; // Data Spacing
		public int[] nppw; // Number of Data Points for Each Pulse Width
		public int gi = 146800; // Group Index
		public short bc = 800; // Backscatter Coefficient
		public int nav; // Number of Averages
		public int at; // Averaging Time --- ??
		public int ar; // Acquision Range
		public int ard; // Acquision Range Distance
		public int fpo = 0; // Front Panel Offset
		public int nf = 40000; // Noise Floor Level
		public int nfsf = 1000; // Noise Floor Scale Factor
		public int po = 0; // Power Offset First Point
		public int lt = 200; // Loss Threshold
		public int rt = 40000; // Reflectance Threshold
		public int et = 3000; // End-of-Fiber Threshold
		public String tt = "ST"; // Trace Type -- ??
		public int wc[] = new int[4]; // Window Coordinates

		public int getSize() {
			return (44 + this.tpw * 10);
		}
	}

	// ----------- Key Events ------------//
	class KeyEvents {
		public short tnke; // Number of Key Events
		public short[] en; // Event Number
		public int[] ept; // Event Propagation Time
		public short[] aci; // Attenuation Coefficient Lead-in-Fiber
		public short[] el; // Event Loss
		public int[] er; // Event Reflectance
		public String[] ec; // Event Code (6 bytes)
		public String[] lmt; // Loss Measurement Event (2bytes)
		public String[] cmt; // Comment
		public int eel = 0; // End-to-End Loss
		public int elmp[] = new int[2]; // End-toEnd Marker Positions
		public int orl; // Optical Return Loss
		public int rlmp[] = new int[2]; // Optical Return Loss Marker Positions

		public int getSize() {
			int size = 26;
			for (int i = 0; i < this.tnke; i++)
				size = size + 22 + this.cmt[i].length() + 1;
			return size;
		}
	}

	// ----------- Link Parameters ------------//
	class LnkParams {
		public short tnl; // Total Number of Landmarks
		public short[] lmn; // Landmark Number
		public String[] lmc; // Landmark Code (2 bytes)
		public int[] lml; // Landmark Location
		public short[] ren; // Related Event Number
		public int[][] gpa; // GPS Info - longitude, latitude (2 ints)
		public short[] fci; // Fiber Correction Factor Lead-in-Fiber
		public int[] smi; // Stealth Marker Entering Landmark
		public int[] sml; // Stealth Marker Leaving Landmark
		public String[] usml; // Units of Stealth Marker Leaving Landmark (2 bytes)
		public short[] mfdl; // Mode Field Diameter Leaving Landmark
		public String[] cmt; // Comment

		public int getSize() {
			int size = 2;
			for (int i = 0; i < this.tnl; i++)
				size = size + 32 + this.cmt[i].length() + 1;
			return size;
		}
	}

	// ----------- Data Points ------------//
	class DataPts {
		public int tndp; // Number of Data Points
		public short tsf; // Total number Scale Factor Used
		public int tps[]; // Total Data Points Using Scale Factor i
		public short sf[]; // Scale Factor i
		public int dsf[][]; // Data

		public int getSize() {
			int size = 6 + this.tsf * 6 + this.tndp * 2;
			// for (int i = 0; i < tsf; i++)
			// size += tps[i] * 2;
			return size;
		}
	}

	// ----------- Checksum ------------//
	class Cksum {
		public int csm; // Checksum

		public int getSize() {
			return 2;
		}
	}

	// ----------- Special Field ------------//
	class Special {
		// public int size; // Size Of Special Field
		public byte[] specData; // Special Data Field

		public int getSize() {
			return this.specData.length;
		}
	}

	public String getOpticalModuleId() {
		return this.supParams.omid;
	}

	public int getPulsewidth() {
		return this.fxdParams.pwu[0];
	}

	public int getAverages() {
		return this.fxdParams.nav;
	}

	public double getBackscatter() {
		return -(double) this.fxdParams.bc / 10d;
	}

	public Date getDate() {
		return new Date(this.fxdParams.dts * 1000);
	}

	public String getUnits() {
		return this.fxdParams.ud;
	}

	public int getWavelength() {
		return this.fxdParams.aw / 10;
	}

	public double getIOR() {
		return this.fxdParams.gi / 100000d;
	}

	/**
	 * Gives resolution (sample spacing) measured in meters.
	 * Generally, gives relative accuracy better than 1e-5.
	 * @return Data resolution, m.
	 */
	public double getResolution() {
		// if BS contains info on data spacing, return that
		if (this.fxdParams.tpw > 0)
		{
			double res = this.fxdParams.ds[0] * 1e-14 * 3e8
				/ (this.fxdParams.gi * 1e-5);
			if (res > 0)
				return res;
		}
		// otherwise, guess it based on total length
		int n = this.dataPts.tndp;
		double res = (this.fxdParams.ar - this.fxdParams.ao) * 3d
			/ ((double) n * (double) this.fxdParams.gi / 1000d);
		return res;
	}

	/**
	 * @return trace range, km
	 */
	public double getRange() {
		return (this.fxdParams.ar - this.fxdParams.ao) * 3d / this.fxdParams.gi;
	}

	public double[] getTraceData() {
		int n = this.dataPts.tndp;
		double[] y = new double[n];

		for (int i = 0; i < this.dataPts.tps[0]; i++)
			y[i] = (65535 - this.dataPts.dsf[0][i]) / 1000d;

		correctReflectogramm1(y);
		shiftTraceByAbsMax(y);
		return y;
	}

	// корректируем мин. значения и первые две точки рефлектограммы
	private void correctReflectogramm1(double[] data) {
		if (data.length < 3)
			return;

		int begin = 300;
		if (begin > data.length / 2)
			begin = data.length / 2;

		double min = data[begin];

		for (int i = begin; i < data.length; i++)
			if (data[i] < min)
				min = data[i];

		if (min != 0)
			for (int i = 0; i < data.length; i++)
				data[i] = data[i] - min;

		for (int i = 0; i <= begin; i++)
			if (data[i] < 0.)
				data[i] = 0.;

		if (data[0] > 0.001)
			data[0] = 0.;

		if (data[1] < 0.001)
			data[1] = data[2] / 2.;
	}

	/**
	 * Выравниваем р/г по абс. максимуму (т.к. считаем его уровнем мертвой зоны).
	 * Это необходимо для нынешнего (june 2005) анализа - по крайней мере для
	 * построения эталона. Код анализа, отображающий р/г на экран,
	 * сам проводит выравнивание, которое, с введением настоящей процедуры
	 * становится излишним.
	 * @todo Привести выравнивание в порядок, одним из трех способов:
	 * <ol>
	 * <li>оставить только настоящее выравнивание, убрав все остальное
	 * выравнивание из анализа как излишнее
	 * (самый прямолинейный и надежный вариант),
	 * <li>сделать унифицированное выравнивание в анализе (фактически,
	 * сделать надстойку над BellcoreStructure и перенести туда
	 * настоящую процедуру),
	 * <li>сделать выравнивание "по месту" - зависящую от того, где она
	 * требуется. Это наиболее гибкий, но наиболее сложный и наименее трудно
	 * контролируемый вариант - однако он может понадобиться, если
	 * выравнивание по мертвой зоне окажется неадекватным - скажем, в случае
	 * отсутствия мертвой зоны.
	 * </ol>
	 */
	private void shiftTraceByAbsMax(double[] data) {
		if (data.length == 0) {
			return;
		}
		// определяем абс. макс.
		double vMax = data[0];
		for (int i = 1; i < data.length; i++) {
			if (vMax < data[i]) {
				vMax = data[i];
			}
		}
		// System.err.println("correctReflectogramm2: vMax = " + vMax);
		// смещаем
		for (int i = 0; i < data.length; i++) {
			data[i] -= vMax;
		}
	}
}

File: hw/trc/BellcoreStructure.cpp
#include "BellcoreStructure.h"
#include <string.h>

const char* BellcoreStructure::MAP_STR = "Map";
const char* BellcoreStructure::GENPARAMS_STR = "GenParams";
const char* BellcoreStructure::SUPPARAMS_STR = "SupParams";
const char* BellcoreStructure::FXDPARAMS_STR = "FxdParams";
const char* BellcoreStructure::KEYEVENTS_STR = "KeyEvents";
const char* BellcoreStructure::LNKPARAMS_STR = "LnkParams";
const char* BellcoreStructure::DATAPOINTS_STR = "DataPts";
const char* BellcoreStructure::SPECIAL_STR = "Special";
const char* BellcoreStructure::CKSUM_STR = "Cksum";

BellcoreStructure::BellcoreStructure() {
	this->hasMap = 0;
	this->hasGen = 0;
	this->hasSup = 0;
	this->hasFxd = 0;
	this->hasKey = 0;
	this->hasLnk = 0;
	this->hasData = 0;
	this->hasSpecial = 0;
	this->specials = 0;
}

BellcoreStructure::~BellcoreStructure() {
	if (this->hasMap)
		delete this->map;
	if (this->hasGen)
		delete this->genParams;
	if (this->hasSup)
		delete this->supParams;
	if (this->hasFxd)
		delete this->fxdParams;
	if (this->hasKey)
		delete this->keyEvents;
	if (this->hasLnk)
		delete this->lnkParams;
	if (this->hasData)
		delete this->dataPts;
	if (this->hasSpecial) {
		for (int i = 0; i < this->specials; i++)
			delete this->special[i];
		delete[] this->special;
	}
}

void BellcoreStructure::addField (unsigned char type) {
	switch (type) {
		case BellcoreStructure::MAP:
			if (!this->hasMap) {
				this->map = new BellcoreStructure::Map();
				this->hasMap = 1;
			}
			break;
		case BellcoreStructure::GENPARAMS:
			if (!this->hasGen) {
				this->genParams = new BellcoreStructure::GenParams();
				this->hasGen = 1;
			}
			break;
		case BellcoreStructure::SUPPARAMS:
			if (!this->hasSup) {
				this->supParams = new BellcoreStructure::SupParams();
				this->hasSup = 1;
			}
			break;
		case BellcoreStructure::FXDPARAMS:
			if (!this->hasFxd) {
				this->fxdParams = new BellcoreStructure::FxdParams();
				this->hasFxd = 1;
			}
			break;
		case BellcoreStructure::KEYEVENTS:
			if (!this->hasKey) {
				this->keyEvents = new BellcoreStructure::KeyEvents();
				this->hasKey = 1;
			}
			break;
		case BellcoreStructure::LNKPARAMS:
			if (!this->hasLnk) {
				this->lnkParams = new BellcoreStructure::LnkParams();
				this->hasLnk = 1;
			}
			break;
		case BellcoreStructure::DATAPOINTS:
			if (!this->hasData) {
				this->dataPts = new BellcoreStructure::DataPts();
				this->hasData = 1;
			}
			break;
		case BellcoreStructure::SPECIAL:
			this->special[this->specials++] = new BellcoreStructure::Special();
			hasSpecial = 1;
			break;
		case BellcoreStructure::CKSUM:
			if (!this->hasMap) {
				this->cksum = new BellcoreStructure::Cksum();
			}
	}
}

unsigned int BellcoreStructure::get_size() const {
	unsigned int size = 0;
	for (int i = 0; i < this->specials; i++)
		size += this->special[i]->get_size();
	if (this->hasMap)
		size += this->map->get_size();
	if (this->hasGen)
		size += this->genParams->get_size();
	if (this->hasSup)
		size += this->supParams->get_size();
	if (this->hasFxd)
		size += this->fxdParams->get_size();
	if (this->hasKey)
		size += this->keyEvents->get_size();
	if (this->hasLnk)
		size += this->lnkParams->get_size();
	if (this->hasData)
		size += this->dataPts->get_size();
	size += this->cksum->get_size();
	return size;
}

BellcoreStructure::Map::Map() {
	this->MRN = 0;
	this->MBS = 0;
	this->NB = 1;
}

BellcoreStructure::Map::~Map() {
	delete[] this->B_id;
	delete[] this->B_rev;
	delete[] this->B_size;
}

unsigned int BellcoreStructure::Map::get_size() const {
	int size = 8;
	for (int i = 1; i < this->NB; i++)
		size = size + strlen(this->B_id[i]) + 7;
	return size;
}

BellcoreStructure::GenParams::GenParams() {
	this->LC = "EN";
	this->CID = " ";
	this->FID = " ";
	this->FT = 652;
	this->NW = 1310;
	this->OL = " ";	
	this->TL = " ";	
	this->CCD = " ";
	this->CDF = "BC";
	this->UO = 0;
	this->UOD = 0;
	this->OP = " ";	
	this->CMT = " ";
}

BellcoreStructure::GenParams::~GenParams() {}

unsigned int BellcoreStructure::GenParams::get_size() const {
	return (10 + 7
			+ strlen(this->CID) + strlen(this->FID)
			+ strlen(this->OL) + strlen(this->TL)// + 6
			+ strlen(this->CCD) + strlen(this->OP) + strlen(this->CMT));
}

BellcoreStructure::SupParams::SupParams() {
	this->SN = " ";
	this->MFID = " ";
	this->OTDR = " ";
	this->OMSN = " ";
	this->SR = " ";
	this->OT = " ";
}

BellcoreStructure::SupParams::~SupParams() {}

unsigned int BellcoreStructure::SupParams::get_size() const {
	return (7 + strlen(this->SN) + strlen(this->MFID)
			+ strlen(this->OTDR) + strlen(this->OMID)
			+ strlen(this->OMSN) + strlen(this->SR) + strlen(this->OT));
}

BellcoreStructure::FxdParams::FxdParams() {
	this->DTS = 0;
	this->UD = "km";
	this->AW = 13100;
	this->AO = 0;
	this->AOD = 0;
	this->TPW = 0;
	this->GI = 146800;
	this->BC = 800;
	this->FPO = 0;
	this->NF = 40000;
	this->NFSF = 1000;
	this->PO = 0;
	this->LT = 200;
	this->RT = 40000;
	this->ET = 3000;
	this->TT = "ST";
	this->WC = new int[4];	//!
}

BellcoreStructure::FxdParams::~FxdParams() {
	delete[] this->WC;
	delete[] this->PWU;
	delete[] this->DS;
	delete[] this->NPPW;
}

unsigned int BellcoreStructure::FxdParams::get_size() const {
	return (44 + TPW * 10);
}

BellcoreStructure::KeyEvents::KeyEvents() {
	this->EEL = 0;
	this->ELMP = new int[2];	//!
	this->RLMP = new int[2];	//!
}

BellcoreStructure::KeyEvents::~KeyEvents() {
	delete[] this->ELMP;
	delete[] this->RLMP;
}

unsigned int BellcoreStructure::KeyEvents::get_size() const {
	int size = 26;
	for (int i = 0; i < this->TNKE; i++)
		size = size + 22 + strlen(this->CMT[i]) + 1;
	return size;
}

BellcoreStructure::LnkParams::LnkParams() {}

BellcoreStructure::LnkParams::~LnkParams() {}

unsigned int BellcoreStructure::LnkParams::get_size() const {
	int size = 2;
	for (int i = 0; i < this->TNL; i++)
		size = size + 32 + strlen(this->CMT[i]) + 1;
	return size;
}

BellcoreStructure::DataPts::DataPts() {}

BellcoreStructure::DataPts::~DataPts() {
	delete[] this->TPS;
	delete[] this->SF;
	short i;
	for (i = 0; i < this->TSF; i++)
		delete[] this->DSF[i];
	delete[] this->DSF;
}

unsigned int BellcoreStructure::DataPts::get_size() const {
	int size = 6 + this->TSF * 6 + this->TNDP * 2;
	//for (int i = 0; i < TSF; i++)
	//		size += TPS[i] * 2;
	return size;
}

BellcoreStructure::Cksum::Cksum() {}

BellcoreStructure::Cksum::~Cksum() {}

unsigned int BellcoreStructure::Cksum::get_size() const {
	return 2;
}

BellcoreStructure::Special::Special() {}

BellcoreStructure::Special::~Special() {}

unsigned int BellcoreStructure::Special::get_size() const {
	return this->size;
}

